### ISSUE TITLE
Automatic node linking to tags and TODO states

### ISSUE THREAD

**Issue by agzam** [org-roam/org-roam#1698]
### Brief Abstract

Would it be possible to automatically link a node when tagged, if there exists [another] node with the same name as the tag?

### Long Description

This comment https://org-roam.discourse.group/t/file-note-todo-states/1821/7?u=agibragimov about TODO states made me think of a similar idea for Org-mode tags.

Since I started using Org-Roam, I find very little use for regular Org-mode tags and TODO states because virtually, they don't make any impact anymore on my workflow.

Instead of TODO states (as I explained in the aforementioned comment), I'd use regular Org-Roam nodes called TODO, DONE, CANCELED, etc. And whenever I need to mark something with the todo state, I'd just add a link to a node with the same name.

That approach is straightforward and has certain benefits: You can see all the TODOs in the backlinks buffer, you can explore them in the graph, etc. 

But that approach is not without cons, e.g., it doesn't prevent you from having multiple [conflicting] todo states in the same node, etc.

So I thought, what if vanilla Org's functionality was preserved, but Org-Roam created those links automatically?

Let's say:

- I'd tag a node with a tag `:meeting:` (or add `#+filetags: meeting` if it's a file-note node)
- Then Org-Roam tries to find a node with the same name
- If found, it automatically sets a link to the node that has `#+title: meeting`

--- 

Of course, with a bit of hacking, one can achieve this exact effect. But the point is not to have an explicit link added (in the text). But only to modify the DB to denote the linkage.

What do you think?

**Comment by jeremyf** [comment]
This is an interesting feature request; I think about my personal website.  Each tag has it's own page, with a list of all blog posts with that tag.  In theory, I can add text for that tag.

The functional idea would be that you now have a node that could be used to describe the tag.  In the above example `:meeting:` could have a description of what you mean by `:meeting:`, maybe some notes about connectivity, even a template for creating a meeting, etc.

I can definitely see a lot of utility in this; I've personally fabricated this functionality by creating links to the anchoring tag in the node's body.

**Comment by jeremyf** [comment]
(I'm rather new to the org-roam event cycle)

I'm thinking that after `org-roam-tag-add`, `org-roam-tag-remove`, saving, and deleting the org-roam file you'd want to update the database.

 @agzam while note quite related, you may be interested in the recent updates from https://github.com/org-roam/org-roam/pull/1693.

**Comment by agzam** [comment]
> The functional idea would be that you now have a node that could be used to describe the tag. In the above example :meeting: could have a description of what you mean by :meeting:, maybe some notes about connectivity, even a template for creating a meeting, etc.

That's right. I'm suggesting to make both - Org-mode tags and todo states first-class citizens of Org-Roam by [automatically] associating them to Org-Roam nodes.

For the tags, the benefits are obvious, but for todo states, it immediately raises some questions. Like:

What if you set the todo state to `TODO` but also manually add a link to a node that is `DONE`? Wouldn't that be confusing?

**Comment by jeremyf** [comment]
I spent some time noodling on the requirements and went down three paths:

1) Adding them to the cache
2) Overriding the `org-roam-backlinks-get` (and other SQL entry points)
3) Create a function that writes a buffer for all nodes that have the given text as the tag (extend this to be the node's title).

## Adding them to the cache

Below are what I believe to be the 4 major features required.  Due to timing issues, I am uncertain when I would expect to add entries to the cache.
```gherkin
Given that I have an org-roam file node with title "meeting"
And given that I have an org-roam node with a tag :meeting
When I rebuild the org-roam cache database
Then there should exist a backlink from the "meeting" titled node to the node tagged :meeting:

Given that the org-roam cache has a backlink (named BL) from "meeting" titled node to the node tagged :meeting:
And I remove the :meeting: tag from the given node
Then the given org-roam cache should not include the given BL backlink

Given that I have an org-roam file node with title "meeting"
When I create an org-roam node with a tag :meeting
Then there should exist a backlink from the "meeting" titled node to the newly created node

Given that I have an org-roam file node with title "meeting"
When I add an org-roam node with a tag :meeting
Then there should exist a backlink from the "meeting" titled node to the newly created node
```

## Overriding the `org-roam-backlinks-get`

This one would remove the "timing" issues and instead inject the functionality at the data retrieval layer.  The `org-roam-backlinks-get` function basically runs the following SQL: `  SELECT source, dest, pos, properties FROM links WHERE dest = $node_id AND type = "id";`  We could create a union that was in essence: `SELECT nodes.id AS source, tags.node_id AS node_id, nodes.pos AS pos, "(outline: \"Unkown\")" AS properties FROM nodes INNER JOIN tags ON WHERE tags.tag = $node_title`

This might give the functionality, but we'd need to make some decisions about the properties.

## Create a function that writes a buffer 

This doesn't manipulate the cache state nor change the data retrieval layer.  Instead, we have a new function that you could call for a buffer to retrieve all nodes with a tag equal to the given string.

## Conclusion

I'm uncertain which path makes sense.  The hardest one to manage is the one relating to state.  The less invasive but easy to be "incorrect" is the overriding.  And the last option (a new function) would not be as integrated.

**Comment by tommyjiang** [comment]
I would also want to vote for this feature. In addition, I would like some feature like select some text and create a node if not exist or linking to an existed node.

Thanks for the excellent work.

**Comment by jeremyf** [comment]
@tommyjiang I recently tried:

1) Mark a region
2) `M-x org-roam-node-insert`

This took the region as the prompted text for finding the node.  And when I hit enter it would find and link or prompt to create a link.  It sounds like you're looking for something that doesn't seek a confirmation prompt.  Is that correct?

**Comment by tommyjiang** [comment]
> @tommyjiang I recently tried:
> 
> 1. Mark a region
> 2. `M-x org-roam-node-insert`
> 
> This took the region as the prompted text for finding the node. And when I hit enter it would find and link or prompt to create a link. It sounds like you're looking for something that doesn't seek a confirmation prompt. Is that correct?

Yes, maybe direct link to the node if exist. Can I use some variable to set this? I do not find such in `org-roam-node-insert` function.

**Comment by jeremyf** [comment]
@tommyjiang Looking over https://github.com/org-roam/org-roam/blob/e9ae19c01cb1fac8256e404b3f9c06f4be5468e6/org-roam-node.el#L567-L606

The following region is what inserts the node:

https://github.com/org-roam/org-roam/blob/e9ae19c01cb1fac8256e404b3f9c06f4be5468e6/org-roam-node.el#L587-L595

But the part that does the lookup uses `org-roam-node-read`

https://github.com/org-roam/org-roam/blob/e9ae19c01cb1fac8256e404b3f9c06f4be5468e6/org-roam-node.el#L435-L463

So you'd need to alter that function?  Or swap out `org-roam-node-insert`'s call to `org-roam-node-read`. I'm uncertain the steps (this is pushing the boundary of my lisp knowledge).  Perhaps an advice function before `org-roam-node-read`?

**Comment by tommyjiang** [comment]
> @tommyjiang Looking over
> 
> https://github.com/org-roam/org-roam/blob/e9ae19c01cb1fac8256e404b3f9c06f4be5468e6/org-roam-node.el#L567-L606
> 
> The following region is what inserts the node:
> 
> https://github.com/org-roam/org-roam/blob/e9ae19c01cb1fac8256e404b3f9c06f4be5468e6/org-roam-node.el#L587-L595
> 
> But the part that does the lookup uses `org-roam-node-read`
> 
> https://github.com/org-roam/org-roam/blob/e9ae19c01cb1fac8256e404b3f9c06f4be5468e6/org-roam-node.el#L435-L463
> 
> So you'd need to alter that function? Or swap out `org-roam-node-insert`'s call to `org-roam-node-read`. I'm uncertain the steps (this is pushing the boundary of my lisp knowledge). Perhaps an advice function before `org-roam-node-read`?

Thanks, I will give a try. Maybe adding a variable to control this behavior would be a solution.

**Comment by sivark** [comment]
I think one simple solution to this would be to have a function which can search for specific regex patterns (eg. inline #foo in logseq) and use the parsed version to create the appropriate node if needed and then link to it. A nicer version might do this for a list of such patterns, each one respectively needing a function to map the variables captured in the regex pattern to the appropriate node/file name.

I would find this extremely useful, but my elisp-fu is unfortunately lacking.

**Comment by martyone** [comment]
I was looking for such a feature too and thinking about it more I realized the proposed approach might not be the right solution to the right problem.

The way tags are handled currently allows partitioning the graph while not adding any visual clutter. This is suitable for recording relations to higher level domains like: computer science, nature, photography, music, sport, work, traveling, ... . It is then possible to narrow node search or limit the portion of graph visible in roam-ui using these tags. At the same time the graph is not full of edges for these relations, which would hurt the signal to noise ratio a lot.

To me it seems the root cause for the existence of the demand for such a feature is that the linking between nodes still feels a bit clumsy for the case when one would like to quickly record a relation in the manner similar to listing keywords.

So I think it should be addressed with a feature that can coexist with the current tags handling.

And I am quite happily using another existing feature for it: aliases. E.g. my node named Emacs has the `#emacs` alias added with `org-roam-alias-add`. When I am capturing a new node I can have e.g. `Relates: #emacs, ...` added on top of it quickly with `org-roam-node-insert` because searching with the leading hash symbol eliminates all the other nodes mentioning Emacs in their title.

For TODO states I think it would be nice if org-roam recorded them as tags similarly to how the org-mode does with the [special properties](https://orgmode.org/manual/Special-Properties.html).
